test/README.md

```diff
@@ -35,3 +35,90 @@ Some examples of test harnesses using these tests, from the source repository:
 - [CST parse/stringify tests](https://github.com/messageformat/messageformat/blob/11c95dab2b25db8454e49ff4daadb817e1d5b770/packages/mf2-messageformat/src/cst/cst.test.ts)
 - [Data model stringify tests](https://github.com/messageformat/messageformat/blob/11c95dab2b25db8454e49ff4daadb817e1d5b770/packages/mf2-messageformat/src/data-model/stringify.test.ts)
 - [Formatting tests](https://github.com/messageformat/messageformat/blob/11c95dab2b25db8454e49ff4daadb817e1d5b770/packages/mf2-messageformat/src/messageformat.test.ts)
+
+## Test Functions
+
+As the behaviour of `:number`, `:datetime`, and other default message functions
+is dependent on locale-specific data and may vary between implementations,
+the following functions are defined for **test use only**:
+
+### `test:func`
+
+This function is valid both as a selector and as a formatter.
+The only option it recognizes is `fd`,
```

Comment:
What does `fd` mean?

We should probably follow the model in the registry.md document in defining these functions. There we have subsections

- operands
- options
- selection
- formatting
- (other stuff)

I think a little bit of design is probably useful here as well. It would be nice to be able to have `:function` (and `:select`) model `:number`'s behavior (with both exact and keyword matching) by defining what numbers match the plural keywords `zero`, `one`, `two`, `few`, `many`, and `other`. That way we can provide tests that implementers can use to see if they match the sorting mechanism in `formatting.md` without writing number/plural selection over.

Another thing to consider is that we don't need options if we can specify how `:function` and `:select` work internally. One thing we could specify is that they'll receive an ordered list or other set of arguments from the test (not in the message) that can be used to program their behavior in a simplistic way, e.g. a regexp that should match the operand.

Reply:
> What does `fd` mean?

It started life as `fractionDigits`, but you asked for this to be distanced from the real `:number` implementation, and I simplified its values to just `0` and `1`, so it makes sense to be a bit more abstract. So now it's just an option that has _some_ effect on the `:test:function` behaviour, and a small set of valid values.

> We should probably follow the model in the registry.md document in defining these functions. There we have subsections

That model doesn't exactly work here, because here we need to be explicit about exactly what this function does, and how it behaves in different settings.

> It would be nice to be able to have `:function` (and `:select`) model `:number`'s behavior (with both exact and keyword matching) by defining what numbers match the plural keywords `zero`, `one`, `two`, `few`, `many`, and `other`. That way we can provide tests that implementers can use to see if they match the sorting mechanism in `formatting.md` without writing number/plural selection over.

That's already supported. Selection on `{1 :test:function}` will only match the `1` key, while selection on `{1 :test:function fd=1}` will match both `1` and `1.0` keys, preferring the latter. This is explicitly different from `:number` selection, again, because you asked for this to be different.

> Another thing to consider is that we don't need options if we can specify how `:function` and `:select` work internally. One thing we could specify is that they'll receive an ordered list or other set of arguments from the test (not in the message) that can be used to program their behavior in a simplistic way, e.g. a regexp that should match the operand.

I would prefer working iteratively, and first seeing where this small family of functions gets us. If/as we discover parts of the spec that do need testing but can't be tested with these functions, then we can reconsider.

I would very much prefer not needing to specify even simple programmable behaviour at the level needed to guarantee exact behaviour across all implementations. For this, regexp is pretty much right out unless there's a ready reference somewhere for a minimum subset that's supported everywhere _and_ for which a validator is available to keep the test data honest.

Reply:
> but you asked for this to be distanced from the real :number implementation

Right. My thinking was that we want something that can be used for conformance testing now and later. We want it to be easy to implement, have deterministic behavior, and cover selection/formatting features required in our spec.

We have [an example](https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md#example-3) in our spec now that should be testable. The example uses `:number`, but replacing it with `:function` could be straightforward.

> That model doesn't exactly work here, because here we need to be explicit about exactly what this function does, and how it behaves in different settings.

That model can work here, because we can describe the operands and options and the proceed to describe specifically how selection works and how formatting works. You're correct that, unlike `:number` or `:datetime`, we need to provide very specific implementation instructions.

Bear in mind that we need to flesh out how to write proposals for the default and "RGI" registry. The test functions should probably go through the motions and use the forms of that process.

> I would very much prefer not needing to specify even simple programmable behaviour at the level needed to guarantee exact behaviour across all implementations. For this, regexp is pretty much right out unless there's a ready reference somewhere for a minimum subset that's supported everywhere and for which a validator is available to keep the test data honest.

Fair enough.

Reply:
> We have [an example](https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md#example-3) in our spec now that should be testable. The example uses `:number`, but replacing it with `:function` could be straightforward.

Sure, that should be pretty directly portable to using `:test:function`. I'll be happy to do so after #767 lands, either within this PR or in a follow-up.

> Bear in mind that we need to flesh out how to write proposals for the default and "RGI" registry. The test functions should probably go through the motions and use the forms of that process.

So could we land this first with its current freeform description, and then update it once the forms of that process are indeed fleshed out? I would really prefer not blocking improvements in tests on figuring out a new proposal process.

Reply:
> So could we land this first with its current freeform description, and then update it once the forms of that process are indeed fleshed out? I would really prefer not blocking improvements in tests on figuring out a new proposal process.

Sure, we can land this. 

However, nothing we do here will effect v45. We have _lots_ to do in the next few months. Each thing we do should not require later additional work to the degree reasonable.

Note that I'm not asking you to wait on a proposal process, but to try and use the flabby outlines of what we have to see what we can learn towards creating said process. If it's too hard, we probably need to look in another direction. FWIW, it would be amusing to see the registry XML or JSON description for `:function`, `:select`, and `:format`

Reply:
> Note that I'm not asking you to wait on a proposal process, but to try and use the flabby outlines of what we have to see what we can learn towards creating said process.

Done.